File: maximum_subarray/maximum_subarray.c
#include <stdio.h>

// gcc maximum_subarray.c -O1 -fsanitize=address -fno-omit-frame-pointer -std=gnu99; ./a.out

int maxSubArray(int* nums, int numsSize){
    /* Given an integer array nums, find the
    * contiguous subarray (containing at least
    * one number) which has the largest sum and
    * return its sum. */
    int previous = nums[0];
    int newSum = 0;
    int i;
    for(i = 0; i < numsSize; i++){
        if(newSum < 0){newSum = 0;}
        newSum += nums[i];
        if(newSum > previous){previous = newSum;}
    }
    return previous;
}


// TESTS
int main(void){
    int returnValue = 0;

    // assert [-2, 1, -3, 4, -1, 2, 1, -5, 4] == 6
    int n[9] = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    int s = sizeof(n) / sizeof(int);
    int result = maxSubArray(n, s);
    printf("\n%d", result);
    if(result != 6){returnValue = 1;}

    // assert [5, 4, -1, 7, 8] == 23
    int nTwo[5] = {5, 4, -1, 7, 8};
    int sTwo = sizeof(nTwo) / sizeof(int);
    int resultTwo = maxSubArray(nTwo, sTwo);
    printf("\n%d", resultTwo);
    if(resultTwo != 23){returnValue = 1;}

    // assert [1] == 1
    int nThree[1] = {1};
    int sThree = sizeof(nThree) / sizeof(int);
    int resultThree = maxSubArray(nThree, sThree);
    printf("\n%d", resultThree);
    if(resultThree != 1){returnValue = 1;}

    printf("\n");
    return returnValue;
}

File: maximum_subarray/maximum_subarray.py
#!/usr/bin/env python3
"""maximum_subarray.py"""
from typing import List
from timeit import timeit


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        """Given an integer array nums, find the
        contiguous subarray (containing at least
        one number) which has the largest sum and
        return its sum."""
        previous = nums[0]
        new_sum = 0
        for num in nums:
            if new_sum < 0:
                new_sum = 0
            new_sum += num
            if new_sum > previous:
                previous = new_sum
        return previous


assert Solution().maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]) == 6
assert Solution().maxSubArray([1]) == 1
assert Solution().maxSubArray([5, 4, -1, 7, 8]) == 23

X = timeit(lambda: Solution().maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]),
           number=1000000)
print(X)
